src/Sylius/Bundle/ThemeBundle/Templating/Locator/TemplateLocator.php

```diff
@@ -0,0 +1,186 @@
+ *
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+class TemplateLocator implements FileLocatorInterface
+{
+    /**
+     * @var ThemeRepositoryInterface
+     */
+    private $themeRepository;
+
+    /**
+     * @var ThemeContextInterface
+     */
+    private $themeContext;
+
+    /**
+     * @var ResourceLocatorInterface
+     */
+    private $bundleResourceLocator;
+
+    /**
+     * @var ResourceLocatorInterface
+     */
+    private $applicationResourceLocator;
+
+    /**
+     * @var array
+     */
+    private $cache;
+
+    /**
+     * @param ThemeRepositoryInterface $themeRepository
+     * @param ThemeContextInterface $themeContext
+     * @param ResourceLocatorInterface $bundleResourceLocator
+     * @param ResourceLocatorInterface $applicationResourceLocator
+     * @param string $cacheDir The cache path
+     */
+    public function __construct(
+        ThemeRepositoryInterface $themeRepository,
+        ThemeContextInterface $themeContext,
+        ResourceLocatorInterface $bundleResourceLocator,
+        ResourceLocatorInterface $applicationResourceLocator,
+        $cacheDir = null
+    ) {
+        $this->themeRepository = $themeRepository;
+        $this->themeContext = $themeContext;
+        $this->bundleResourceLocator = $bundleResourceLocator;
+        $this->applicationResourceLocator = $applicationResourceLocator;
+
+        if (null !== $cacheDir && is_file($cache = $cacheDir . '/templates.php')) {
+            $this->cache = require $cache;
+        }
+    }
+
+    /**
+     * Returns a full path for a given file.
+     *
+     * @param TemplateReferenceInterface $template A template
+     * @param string $currentPath Unused
+     * @param bool $first Unused
+     *
+     * @return string The full path for the file
+     *
+     * @throws \InvalidArgumentException When the template is not an instance of TemplateReferenceInterface
+     * @throws \InvalidArgumentException When the template file can not be found
+     */
+    public function locate($template, $currentPath = null, $first = true)
+    {
+        if (!$template instanceof TemplateReferenceInterface) {
+            throw new \InvalidArgumentException("The template must be an instance of TemplateReferenceInterface.");
+        }
+
+        $themes = $this->themeContext->getThemes();
+
+        $templatePath = $this->locateTemplateUsingThemes($template, $themes);
+
+        if (null === $templatePath) {
+            throw new \InvalidArgumentException(
+                sprintf('Unable to find template "%s" (using themes: %s).', $template, join(', ', $themes))
+            );
+        }
+
+        return $templatePath;
+    }
+
+    /**
+     * @param TemplateReferenceInterface $template
+     * @param ThemeInterface[] $themes
+     *
+     * @return null|string
+     */
+    protected function locateTemplateUsingThemes(TemplateReferenceInterface $template, array $themes)
+    {
+        foreach (array_merge($themes, [null]) as $theme) {
```

Comment:
To be honest, I have no idea (I haven't used it yet since I had written it). In the next weeks I'll be doing a project based on themes, so I'll dive into ThemeBundle and try to make it better :)

The first ideas are to allow themes to be loaded at the runtime, remove (or extract) unnecessary state in services (like in `Locator`) and to make templates and assets storage-independent.